prototypes/matrix_free_navier_stokes/matrix_free_navier_stokes.cc

```diff
@@ -0,0 +1,2420 @@
+        }
+    }
+}
+
+template <int dim, typename number>
+types::global_dof_index
+NavierStokesOperator<dim, number>::m() const
+{
+  if (this->matrix_free.get_mg_level() != numbers::invalid_unsigned_int)
+    return this->matrix_free.get_dof_handler().n_dofs(
+      this->matrix_free.get_mg_level());
+  else
+    return this->matrix_free.get_dof_handler().n_dofs();
+}
+
+template <int dim, typename number>
+number
+NavierStokesOperator<dim, number>::el(unsigned int, unsigned int) const
+{
+  Assert(false, ExcNotImplemented());
+  return 0;
+}
+
+template <int dim, typename number>
+void
+NavierStokesOperator<dim, number>::clear()
+{}
+
+template <int dim, typename number>
+void
+NavierStokesOperator<dim, number>::initialize_dof_vector(VectorType &vec) const
+{
+  matrix_free.initialize_dof_vector(vec);
+}
+
+template <int dim, typename number>
+const std::shared_ptr<const Utilities::MPI::Partitioner> &
+NavierStokesOperator<dim, number>::get_vector_partitioner() const
+{
+  return matrix_free.get_vector_partitioner();
+}
+
+// Performs an operator evaluation by looping over all cells
+// and evaluating the integrals in the matrix-free way
+template <int dim, typename number>
+void
+NavierStokesOperator<dim, number>::vmult(VectorType &      dst,
+                                         const VectorType &src) const
+{
+  // save values for edge constrained dofs and set them to 0 in src vector
+  for (unsigned int i = 0; i < edge_constrained_indices.size(); ++i)
+    {
+      edge_constrained_values[i] = std::pair<number, number>(
+        src.local_element(edge_constrained_indices[i]),
+        dst.local_element(edge_constrained_indices[i]));
+
+      const_cast<LinearAlgebra::distributed::Vector<number> &>(src)
+        .local_element(edge_constrained_indices[i]) = 0.;
+    }
+
+  this->matrix_free.cell_loop(
+    &NavierStokesOperator::do_cell_integral_range, this, dst, src, true);
+
+  // set constrained dofs as the sum of current dst value and src value
+  for (unsigned int i = 0; i < constrained_indices.size(); ++i)
+    dst.local_element(constrained_indices[i]) =
+      src.local_element(constrained_indices[i]);
+
+  // restoring edge constrained dofs in src and dst
+  for (unsigned int i = 0; i < edge_constrained_indices.size(); ++i)
+    {
+      const_cast<LinearAlgebra::distributed::Vector<number> &>(src)
+        .local_element(edge_constrained_indices[i]) =
+        edge_constrained_values[i].first;
+      dst.local_element(edge_constrained_indices[i]) =
+        edge_constrained_values[i].first;
+    }
+}
+
+// Performs the transposed operator evaluation. Since we have
+// non-symmetric matrices this is different from the vmult call.
+// TODO: implement this correctly (let's start with something symmetric)
+template <int dim, typename number>
+void
+NavierStokesOperator<dim, number>::Tvmult(VectorType &      dst,
+                                          const VectorType &src) const
+{
```

Comment:
Not sure how this impact the resolution process. Can it have a significant impact to consider only a symmetric matrice?

Reply:
We are not using the transposed evaluation of the operator at the moment. So it does not have any impact right now. I will write a comment on this :)